packages/react/src/forwardRef.tsx

```diff
@@ -15,20 +15,28 @@ export type PropsOf<T extends ElementType> = ComponentPropsWithoutRef<T> & {
   as?: ElementType
 }
 
+/**
+ * Assign property types from right to left.
+ * Think `Object.assign` for types.
+ *
+ * @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign
+ */
+export type Assign<Target, Source> = Omit<Target, keyof Source> & Source
+
 export type OmitCommonProps<
   Target,
   OmitAdditionalProps extends string | number | symbol = never,
 > = Omit<Target, 'transition' | 'as' | 'color' | OmitAdditionalProps>
 
-export type AssignCommon<
-  SourceProps extends object = Record<string, unknown>,
-  OverrideProps extends object = Record<string, unknown>,
-> = OmitCommonProps<SourceProps, keyof OverrideProps> & OverrideProps
+type AssignCommon<
+  SourceProps extends object = Record<never, never>,
+  OverrideProps extends object = Record<never, never>,
+> = Assign<OmitCommonProps<SourceProps>, OverrideProps>
 
-export type MergeWithAs<
-  ComponentProps extends Record<string, unknown>,
-  AsProps extends Record<string, unknown>,
-  AdditionalProps extends Record<string, unknown> = Record<string, unknown>,
+type MergeWithAs<
+  ComponentProps extends object,
+  AsProps extends object,
+  AdditionalProps extends object = Record<never, never>,
```

Comment:
I'm wondering if we need to do this:

```suggestion
  AdditionalProps extends Record<string, unknown> = Record<string, unknown>,
```

Reply:
Sounds valid, I'll check it shortly! But the second part `Record<never, never>` is necessary for the autocomplete to show up - `Record<string, unknown>` allows all props and does not throw any errors.

Reply:
✅